clients/include/testing_gebsr2csr.hpp

```diff
@@ -0,0 +1,626 @@
+        auto dcsr_col_ind_managed
+            = hipsparse_unique_ptr{device_malloc(sizeof(int) * safe_size), device_free};
+        auto dcsr_val_managed
+            = hipsparse_unique_ptr{device_malloc(sizeof(T) * safe_size), device_free};
+
+        int* dbsr_row_ptr = (int*)dbsr_row_ptr_managed.get();
+        int* dbsr_col_ind = (int*)dbsr_col_ind_managed.get();
+        T*   dbsr_val     = (T*)dbsr_val_managed.get();
+        int* dcsr_row_ptr = (int*)dcsr_row_ptr_managed.get();
+        int* dcsr_col_ind = (int*)dcsr_col_ind_managed.get();
+        T*   dcsr_val     = (T*)dcsr_val_managed.get();
+
+        if(!dbsr_row_ptr || !dbsr_col_ind || !dbsr_val || !dcsr_row_ptr || !dcsr_col_ind
+           || !dcsr_val)
+        {
+            verify_hipsparse_status_success(HIPSPARSE_STATUS_ALLOC_FAILED,
+                                            "!dbsr_row_ptr || !dbsr_col_ind || !dbsr_val || "
+                                            "!dcsr_row_ptr || !dcsr_col_ind || !dcsr_val");
+            return HIPSPARSE_STATUS_ALLOC_FAILED;
+        }
+
+        status = hipsparseXgebsr2csr(handle,
+                                     dir,
+                                     mb,
+                                     nb,
+                                     bsr_descr,
+                                     dbsr_val,
+                                     dbsr_row_ptr,
+                                     dbsr_col_ind,
+                                     row_block_dim,
+                                     col_block_dim,
+                                     csr_descr,
+                                     dcsr_val,
+                                     dcsr_row_ptr,
+                                     dcsr_col_ind);
+
+        if(mb < 0 || nb < 0 || row_block_dim < 0 || col_block_dim < 0)
+        {
+            verify_hipsparse_status_invalid_size(
+                status, "Error: mb < 0 || nb < 0 || row_block_dim < 0 || col_block_dim < 0");
+        }
+        else
+        {
+            verify_hipsparse_status_success(
+                status, "mb >= 0 && nb >= 0 && row_block_dim >= 0 && col_block_dim >= 0");
+        }
+
+        return HIPSPARSE_STATUS_SUCCESS;
+    }
+
+    // Read or construct CSR matrix
+    std::vector<int> bsr_row_ptr;
+    std::vector<int> bsr_col_ind;
+    std::vector<T>   bsr_val;
+    int              nnzb;
+    srand(12345ULL);
+    if(binfile != "")
+    {
+        if(read_bin_matrix(
+               binfile.c_str(), mb, nb, nnzb, bsr_row_ptr, bsr_col_ind, bsr_val, bsr_idx_base)
+           != 0)
+        {
+            fprintf(stderr, "Cannot open [read] %s\n", binfile.c_str());
+            return HIPSPARSE_STATUS_INTERNAL_ERROR;
+        }
+    }
+    else if(argus.laplacian)
+    {
+        mb = nb
+            = gen_2d_laplacian(argus.laplacian, bsr_row_ptr, bsr_col_ind, bsr_val, bsr_idx_base);
+        nnzb = bsr_row_ptr[m];
```

Comment:
nnzb = bsr_row_ptr[mb]?

Reply:
I would actually use nnz not nnzb as this becomes confusing. I realize you are constructing a csr matrix of size mb by nb instead of m by n so that you can later construct the gebsr matrix by just modifying the values array but I think while the bsr_row_ptr etc arrays are actually representing a CSR matrix, you should use nnz not nnzb.

Reply:
good catch.